documentation/specs/proposed/Wasm-tasks.md

```diff
@@ -0,0 +1,344 @@
+3. write your Rust program
+3. `project.csproj`
+```xml
+  <Target Name="CompileAndRun" BeforeTargets="Build">
+    <Exec Command="cargo build --target wasm32-wasi --release">
+    <Exec Command="wasmtime run path_to_compiled_rust_program.wasm" />
+</Target>
+```
+4. `dotnet build`
+
+This is quite cumbersome and does not provide a way to pass parameters to the "task" or get outputs from it.
+
+## Goals for the Wasm tasks feature
+1. specify how a Wasm/WASI task should communicate with MSBuild, and what it should contain to be recognized as a task
+2. Write an `ITaskFactory` and a supporting `ITask` classes that when passed a `.wasm` file implementing required functions runs it as an MSBuild task
+3. Demos/examples
+
+### Prototype features
+Prototypes are implemented in [https://github.com/JanProvaznik/MSBuildWasm](https://github.com/JanProvaznik/MSBuildWasm)
+- [ ] WasmExec class taking a .wasm file as a parameter - just runs the file with Wasmtime 
+    - nudges the user to parametrize access to resources
+- [ ] WasmTask - creating tasks from .wasm files
+    - [x] Specification for what should this .wasm file export and how it will be ran
+    - [ ] ITaskFactory that let's msbuild use task parameters defined inside the .wasm module
+- [ ] Rust example
+#### User Experience
+1. The user Writes a task in Rust based on the template.
+2. The user adds the task to their .proj file and it runs and logs as if it were a C# task. 
+```xml
+<UsingTask TaskName="FancyWasmTask"
+           AssemblyFile="path/MSBuildWasm.dll"
+           TaskFactory="WasmTaskFactory">
+  <Task>
+    <WasmModule>compiled_task_implementation.wasm</WasmModule>
+  </Task>
+</UsingTask>
+
+<Target Name="name">
+<FancyWasmTask Param="..." Param2="asdf">
+<Output .../>
+</FancyWasiTask>
+</Target>
+```
+
+### Advanced features
+- [ ] ~~.NET example~~ (WASIp1 will not be supported in .NET)
+- [ ] integrating pipeline for creating Wasm/WASI tasks from code in Rust 
+    - [ ] investigate integrating tools compiling languages to Wasm/WASI
+    - On task level
+        - [ ] RustTaskFactory
+        - exploring other languages (Go, C/C++, Zig)
+- [x] investigate running an arbitrary .NET task distributed as a dll in the WASI sandbox (👀 Mono runtime)
+    - Due to implementing WasmTasks with WASIp1, it will not compatible with .NET runtime effort to support WASIp2
+
+
+## Design
+### diagram
+
+```mermaid
+flowchart TD
+    A[MSBuild] -->|Evaluation| B[WasmTaskFactory]
+    A -->|Target execution| C[TaskExecutionHost]
+    C -->|instantiate and\n set parameters from XML| D[WasmTask]
+    H[Rust/C#/Go] -->|"compile using wasi-sdk"| G
+    D -->|gather output \nfor use in other tasks| C 
+    D -->|execute| E[wasmtime-dotnet]
+    E <--> F[Wasmtime]
+
+    B -->|Create Type for a specific WasmTask| D
+    B -->|read what the task expects as parameters| E
+    B -->|save path to task parameters| G[.wasm module]
```

Comment:
What do the arrows represent in this diagram? I thought it was "controls" or inputs but now I don't think I understand.

Reply:
How calls/information flow. I reworded it and added ordering.